tests/usbprotocol/test_comms.py

```diff
@@ -0,0 +1,97 @@
+import random
+import unittest
+
+from panda import LEN_TO_DLC, DLC_TO_LEN, pack_can_buffer, unpack_can_buffer
+from panda.tests.usbprotocol.libpandaprotocol_py import ffi, libpandaprotocol as lpp
+
+TX_QUEUES = (lpp.tx1_q, lpp.tx2_q, lpp.tx3_q, lpp.txgmlan_q)
+
+def package_can_msg(msg):
+  addr, _, dat, bus = msg
+  ret = ffi.new('CANPacket_t *')
+  ret[0].extended = 1 if addr >= 0x800 else 0
+  ret[0].addr = addr
+  ret[0].data_len_code = LEN_TO_DLC[len(dat)]
+  ret[0].bus = bus
+  ret[0].data = bytes(dat)
+  return ret
+
+def unpackage_can_msg(pkt):
+  dat_len = DLC_TO_LEN[pkt[0].data_len_code]
+  dat = bytes(pkt[0].data[0:dat_len])
+  return pkt[0].addr, 0, dat, pkt[0].bus
+
+class PandaCommsTest(unittest.TestCase):
```

Comment:
do we really need these new tests? can we cover this more generically with a better loopback test?

Reply:
These were very helpful in finding edge cases, and are a lot easier to run and debug with than HITL. In the future, I can see these being useful in combination with boardd tests to test the whole chain without needing hardware